pkg/applier/unstructured.go

```diff
@@ -0,0 +1,68 @@
+/*
+Copyright 2022 k0s authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package applier
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes/scheme"
+)
+
+// ToUnstructured converts the given runtime object to an unstructured one using
+// the given scheme. The scheme can be nil, in which case client-go's default
+// scheme is used.
+func ToUnstructured(s *runtime.Scheme, object runtime.Object) (u *unstructured.Unstructured, err error) {
+	u = new(unstructured.Unstructured)
+	u.Object, err = runtime.DefaultUnstructuredConverter.ToUnstructured(object)
+	if err != nil {
+		return nil, err
+	}
+
+	gvk := u.GroupVersionKind()
+	if gvk.Group == "" || gvk.Kind == "" {
+		if s == nil {
+			s = scheme.Scheme
+		}
+
+		kinds, _, err := s.ObjectKinds(object)
+		if err != nil {
+			return nil, err
+		}
+		apiVersion, kind := kinds[0].ToAPIVersionAndKind()
```

Comment:
I'm not quite sure if I'm right, but I think s.ObjectKinds called in line 43 return the apiVersion as well. This would make calling this function redundant and if I'm not mistaken this would do the same:

```
		kinds, apiVersion, err := s.ObjectKinds(object)
		if err != nil {
			return nil, err
		}
		u.SetAPIVersion(apiVersion)
		u.SetKind(kinds[0])
	}

	return u, nil
```

Again I want to stress that I'm not 100% certain about this and I haven't tested it myself.

Other than this, the code looks good to me.

Reply:
[runtime#Scheme.ObjectKinds](https://pkg.go.dev/k8s.io/apimachinery@v0.25.4/pkg/runtime#Scheme.ObjectKinds) returns `([]schema.GroupVersionKind, bool, error)`, basically a slice of possible GVKs.

Reply:
Ugh I should have checked better, sorry.